src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala

```diff
@@ -434,6 +439,103 @@ trait BronzeTransforms extends SparkSessionWrapper {
 
   }
 
+  private def landClusterEvents(clusterIDs:Array[String],
+                                startTime: TimeTypes,
+                                endTime: TimeTypes,
+                                apiEnv: ApiEnv,
+                                tmpClusterEventsSuccessPath:String,
+                                tmpClusterEventsErrorPath:String)={
+    val finalResponseCount = clusterIDs.length
+    var responseCounter = 0
+    var apiResponseArray = Collections.synchronizedList(new util.ArrayList[String]())
+    var apiErrorArray = Collections.synchronizedList(new util.ArrayList[String]())
+    implicit val ec: ExecutionContextExecutor = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(apiEnv.threadPoolSize))
+    //TODO identify the best practice to implement the future.
+    for (i <- clusterIDs.indices) {
+      val jsonQuery = s"""{"cluster_id":"${clusterIDs(i)}","start_time":${startTime.asUnixTimeMilli},"end_time":${endTime.asUnixTimeMilli},"limit":500}"""
+      val future = Future {
+        val apiObj = ApiCallV2(apiEnv, "clusters/events", jsonQuery, tmpClusterEventsSuccessPath).executeMultiThread()
+        synchronized {
+          apiResponseArray.addAll(apiObj)
+          if (apiResponseArray.size() >= apiEnv.successBatchSize) {
+            Helpers.writeMicroBatchToTempLocation(tmpClusterEventsSuccessPath, apiResponseArray.toString)
+            apiResponseArray = Collections.synchronizedList(new util.ArrayList[String]())
+          }
+        }
+
+      }
+      future.onComplete {
+        case Success(_) =>
+          responseCounter = responseCounter + 1
+
+        case Failure(e) =>
+          if (e.isInstanceOf[ApiCallFailureV2]) {
+            synchronized {
+              apiErrorArray.add(e.getMessage)
+              if (apiErrorArray.size() >= apiEnv.errorBatchSize) {
+                Helpers.writeMicroBatchToTempLocation(tmpClusterEventsErrorPath, apiErrorArray.toString)
+                apiErrorArray = Collections.synchronizedList(new util.ArrayList[String]())
+              }
+            }
+
+          }
+          responseCounter = responseCounter + 1
+          logger.log(Level.ERROR, "Future failure message: " + e.getMessage, e)
+      }
+    }
+    while (responseCounter < finalResponseCount) {
```

Comment:
As I feared, this can lock up the module -- when there are failures in some of the clusters -- such as deleted clusters or other errors -- the responses are never captured so it sleeps forever waiting for those hanging threads to complete.

At the very least -- for now, we must have a timeout and make sure those get sent to error
Better would be to make sure the response accounts for successes + errors
Best (for later) identify better logic control to wait for futures to complete -- as already discussed.

Reply:
i guess we will always get a response even its a failure, in case of the cluster is not found it will send an api exception which we take into count as a response.

for now i am implementing timeout of 5 minutes.If we dont get any new response even waiting for 5 mints it will throw an exception.